lib/http/request/writer.rb

```diff
@@ -51,12 +51,16 @@ def connect_through_proxy
       # Adds the headers to the header array for the given request body we are working
       # with
       def add_body_type_headers
-        if @body.is_a?(String) && !@headers[Headers::CONTENT_LENGTH]
-          @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.bytesize}"
-        elsif @body.nil? && !@headers[Headers::CONTENT_LENGTH]
-          @request_header << "#{Headers::CONTENT_LENGTH}: 0"
-        elsif @body.is_a?(Enumerable) && CHUNKED != @headers[Headers::TRANSFER_ENCODING]
-          raise(RequestError, "invalid transfer encoding")
+        unless @headers[Headers::CONTENT_LENGTH]
+          if @body.is_a?(String)
+            @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.bytesize}"
+          elsif @body.respond_to?(:read)
+            @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.size}"
```

Comment:
The reason why I wasn't checking here whether the IO object responds to `size` was because I wanted that it's *required* for the IO object to respond to `size` if `Content-Length` wasn't passed in, because for non-chunked requests `Content-Length` header is required (I would document this in the wiki).